presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java

```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
+    {
+        List<CassandraColumnHandle> clusteringKeyColumns = cassandraTable.getClusteringKeyColumns();
+        Set<List<Object>> clusteringKeysSet = getClusteringKeysSet(cassandraTable, constraint.getSummary());
+        if (clusteringKeysSet == null || clusteringKeysSet.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // todo Is it correct the way predicates are being appended to each other?
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int j = 0; j < clusteringKeysSet.size(); j++) {
+            List<Object> clusteringKeys = clusteringKeysSet.iterator().next();
+
+            for (int i = 0; i < clusteringKeys.size(); i++) {
+                if ((i > 0) || (j > 0)) {
+                    stringBuilder.append(" AND ");
+                }
+
+                stringBuilder.append(CassandraCqlUtils.validColumnName(clusteringKeyColumns.get(i).getName()));
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(clusteringKeys.get(i)), clusteringKeyColumns.get(i).getCassandraType()));
+            }
+        }
+        return Optional.of(stringBuilder.toString());
+    }
+
+    private static Set<List<Object>> getClusteringKeysSet(CassandraTable table, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        boolean predicateMissing = false;
+        ImmutableList.Builder<Set<Object>> clusteringColumnValues = ImmutableList.builder();
+        for (CassandraColumnHandle columnHandle : table.getClusteringKeyColumns()) {
+            Domain domain = tupleDomain.getDomains().get().get(columnHandle);
+
+            if (domain == null) {
+                predicateMissing = true; // acknowledge that predicate on a clustering key is absent
+                continue;
+            }
+            else if (domain != null && predicateMissing) { // clustering keys predicate is not a prefix
+                return ImmutableSet.of();
+            }
+
+            // todo does cassandra allow null clustering keys?
+            if (domain.isNullAllowed()) {
+                return ImmutableSet.of();
+            }
+
+            Set<Object> values = domain.getValues().getValuesProcessor().transform(
+                    ranges -> {
+                        ImmutableSet.Builder<Object> columnValues = ImmutableSet.builder();
+                        for (Range range : ranges.getOrderedRanges()) {
+                            // Prior to Cassandra 2.2, only the last clustering column can have a range of values
+                            if (!range.isSingleValue() && !table.isLastClusteringKeyColumn(columnHandle)) {
+                                return ImmutableSet.of();
+                            }
+                            else if (!range.isSingleValue() && table.isLastClusteringKeyColumn(columnHandle)) {
```

Comment:
Remove this `isLastClusteringKeyColumn` method. Use just `for(;;)` loop. And determine whether it is last clustering key by the index.